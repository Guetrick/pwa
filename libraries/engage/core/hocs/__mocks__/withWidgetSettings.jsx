import React from 'react';

/**
 * Mock for the withWidgetSettings HOC.
 * @param {Function} WrappedComponent The react component to wrap.
 * @returns {JSX}
 */
export function withWidgetSettings(WrappedComponent) {
  const settings = {
    mocked: 'widgetSetting',
  };

  return props => (<WrappedComponent widgetSettings={settings} {...props} />);
}
