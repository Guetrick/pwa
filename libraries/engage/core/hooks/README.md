# Shopgate ENGAGE React Hooks

This library provides you with a collecton of usefull React Hooks that you can use to build out new features in Shopgate ENGAGE.

## Content

### `useCustomRoute`

This Hook returns all necessary React elements that you need to build a custom route:

* `View` - The view component that you should wrap around your view content
* `Appbar` - The app bar component for the appropriate theme (GMD or iOS),
* `params` - The current Route's parameters.

#### Usage

```jsx
import { useCustomRoute, Route } from '@shopgate/engage/core';

function MyCustomView() {
  const { View, AppBar, params } = useCustomRoute();

  return (
    <View>
      <AppBar title="My Custom View" />
      {/* ... your custom content goes here */}
    </View>
  )
}

function MyCustomRoute() {
  return (
    <Route pattern="/my-custom-url" component={MyCustomView} />
  );
}

export default MyCustomRoute;
```
