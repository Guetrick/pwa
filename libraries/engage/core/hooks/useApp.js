import { useContext } from 'react';
import AppContext from '../contexts/AppContext';

/**
 * Provides the properties of the app context.
 * @returns {Object}
 */
export function useApp() {
  const appContext = useContext(AppContext);
  return appContext;
}
