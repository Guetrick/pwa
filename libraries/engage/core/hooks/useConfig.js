import appConfig from '@shopgate/pwa-common/helpers/config';

/**
 * Provides the global theme configuration.
 * @returns {Object}
 */
export function useConfig() {
  const { theme: { pages, ...config } = {} } = appConfig;
  return config || {};
}
