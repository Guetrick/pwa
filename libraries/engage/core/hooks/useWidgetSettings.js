import { useRoute } from './useRoute';
import { getWidgetSettings } from '../config/getWidgetSettings';

/**
 * Retrieves the settings of a specific widget by its id and an optional index on the current page.
 * If no index is given it will return the first found widget settings by widgetId on that page.
 * Widget settings that are not defined in the widget settings directly, are inherited from page
 * and theme settings in the upper hierarchy of the config, scoped by the widgetId as key.
 *
 * @param {string} widgetId The id of the widget to look for, which must exist in the config.
 * @param {number|undefined} [index=0] The optional index of the widget.
 * @returns {Object}
 */
export function useWidgetSettings(widgetId, index = 0) {
  const { pattern: pagePattern } = useRoute();
  return getWidgetSettings(pagePattern, widgetId, index);
}
