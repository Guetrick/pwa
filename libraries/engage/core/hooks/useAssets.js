import { useConfig } from './useConfig';

/**
 * Retrieves the global theme assets.
 * @returns {Object}
 */
export function useAssets() {
  const { assets = {} } = useConfig();

  return assets;
}
