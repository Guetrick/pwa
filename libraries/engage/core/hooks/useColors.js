import { useConfig } from './useConfig';

/**
 * Retrieves the global theme colors.
 * @returns {Object}
 */
export function useColors() {
  const { colors = {} } = useConfig();

  return colors;
}
