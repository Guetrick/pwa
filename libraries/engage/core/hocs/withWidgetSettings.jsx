import React from 'react';
import { useWidgetSettings } from '../hooks/useWidgetSettings';

/**
 * Injects the widget - configuration - settings for a specific widget by its id and
 * an optional index. into the desired component.  If no index is given it will inject the first
 * found widget - configuration - settings by widgetId.
 * @param {Function} WrappedComponent The react component to wrap.
 * @param {string} widgetId The id of the widget to look for, which must exist in the config.
 * @param {number|undefined} [index] The optional index of the widget.
 * @returns {Object}
 */
export function withWidgetSettings(WrappedComponent, widgetId, index) {
  return (props) => {
    const settings = useWidgetSettings(widgetId, index);

    return (<WrappedComponent widgetSettings={settings} {...props} />);
  };
}
