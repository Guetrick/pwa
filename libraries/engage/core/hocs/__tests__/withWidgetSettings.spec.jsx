import React from 'react';
import { mount } from 'enzyme';
import { withWidgetSettings } from '../withWidgetSettings';
import { useWidgetSettings } from '../../hooks/useWidgetSettings';

jest.mock('../../hooks/useWidgetSettings', () => ({
  useWidgetSettings: jest.fn(),
}));

// eslint-disable-next-line require-jsdoc
const MockComponent = () => null;

const widgetSettings = {
  some: 'setting',
};

describe('engage > core > hocs > withWidgetSettings', () => {
  beforeEach(() => {
    useWidgetSettings.mockReturnValue(widgetSettings);
  });

  it('should merge add the widget settings to a wrapped component', () => {
    const ComposedComponent = withWidgetSettings(MockComponent);
    const wrapper = mount(<ComposedComponent />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(MockComponent).prop('widgetSettings')).toEqual(widgetSettings);
  });
});
