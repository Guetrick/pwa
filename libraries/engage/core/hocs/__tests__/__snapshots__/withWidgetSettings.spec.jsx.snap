// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`engage > core > hocs > withWidgetSettings should merge add the widget settings to a wrapped component 1`] = `
<Component
  widgetSettings={
    Object {
      "some": "setting",
    }
  }
/>
`;
